add array_windows / locus_windows, refactor ld_prune, remove filteredEntriesTable and UpperIndexBounds (v2)

jbloom22 opened the pull request:
This is a rebased update to the closed PR #3715. Updates:

- `locus_table` is now `locus_expr`, an expression on a table or matrix table. This is more flexible, and avoids unnatural requirements on key (rather than order).

- uses `global_position` and requires/checks ascending order, rather than reordering within the method. Re-ordering, if non-trivial, would silently invalidate the results with respect to the row-order of the source. This also avoids a potentially unnecessary shuffle, and addresses the issue that contig order may not be alphabetical in the reference.

- keeps the size of data collected close to the minimal data necessary (in particular, no collection of contigs). When `coord_expr` is not set, its an array of int. When `coord_expr` is set, its an array of (int, float). A tuple of arrays would be better, but directly collecting two arrays in order would require two actions with the current infrastructure since agg.collect() does not preserve order.

jbloom22 commented:
@liameabbott I think you should go ahead and merge #3859. Once this is in, you can then use `locus_windows` to simplify, reduce memory req, and be more robust to catching out-of-order loci.

tpoterba commented on python/hail/docs/methods/misc.rst:
```diff
@@ -11,7 +11,9 @@ Miscellaneous
     grep
     maximal_independent_set
     rename_duplicates
+    array_windows
```
this doesn't seem broadly useful enough to warrant being in the `hl.` namespace

tpoterba replied:
my suggestion is that these should go in experimental

tpoterba commented on python/hail/docs/methods/genetics.rst:
```diff
@@ -19,6 +19,7 @@ Genetics
     realized_relationship_matrix
     impute_sex
     ld_prune
+    locus_windows
```
same comment as below

tpoterba commented on python/hail/methods/misc.py:
```diff
@@ -358,4 +368,216 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+    check_row_indexed('locus_windows', locus_expr)
+
+    use_pos = coord_expr is None
+    src = locus_expr._indices.source
```
unused?

jbloom22 commented:
I'd argue that it's broadly useful, but rather the issue is that it's useful at a lower level of abstraction (e.g. how it's used in `ld_prune`, composed with `sparsify_row_intervals`).

So I see why the `hl` namespace is may be too high level, but it's also strange to put in experimental a function that is used in non-experimental (as well as experimental) methods. One option is to underscore the method for use outside experimental, but expose through the experimental module. Another is to put it in a submodule, like genetics.

I'd like to expose more high-level applications directly (e.g. an `ld_matrix` function that takes an optional `radius` and `coord_expr` and returns the sparse block matrix), and we can think about re-implementing in terms of scans once they come online (deriving the stops from the starts) and zipping the intervals and blocks without ever localizing should memory or performance be an issue.
I don't want to hide it entirely in the meantime as several groups want to make use of it already.

jbloom22 commented:
@jigold I assigned to you since this essentially re-opens #3715 with your comments addressed. After playing around, felt that it didn't make sense at this point to design flexible ndarray checking at the level of typecheckers. The requirements vary quite a bit between these two functions. If you feel strongly, let's discuss. I'll leave Tim's review un-dismissed until we've agreed where to put these functions, but no need to wait on it w.r.t reviewing (once you're back).

jigold commented on python/hail/tests/test_methods.py:
```diff
@@ -1360,26 +1362,34 @@ def test_ld_prune(self):
-        ds = hl.balding_nichols_model(n_populations=1, n_samples=10, n_variants=100)
-        pruned_table = hl.ld_prune(ds.GT, r2=0.1, bp_window_size=0)
+        ds = hl.balding_nichols_model(n_populations=1, n_samples=10, n_variants=10)
+        pruned_table = hl.ld_prune(ds.GT, r2=0.0, bp_window_size=0)
```
Isn't this the opposite of no pruning? I would have thought the R^2 threshold would be 1.0.

jbloom22 replied:
`r2=1.0` could still resulting in pruning if there were perfectly correlated variants.

By setting `r2=0.0` I'm being more explicit that no pruning results with `bp_window_size=0` on a matrix table with strictly increasing variant position.

jigold commented on src/main/scala/is/hail/methods/UpperIndexBounds.scala:
```diff
@@ -1,71 +0,0 @@
-import org.apache.spark.sql.Row
-
-object UpperIndexBounds {
-  def groupPositionsByKey(t: Table): RDD[Array[Int]] = {
```
I'm deferring to you that all of the scala code can be deleted.

tpoterba commented on python/hail/methods/misc.py:
```diff
@@ -358,4 +368,213 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+
+    Examples
+    --------
+    Windows with 2bp radius for one contig with positions 1, 2, 3, 4, 5:
```
should we vote on which we prefer? The extra newline is not necessary for the file to build correctly, and I think it adds unnecessary spacing

jigold replied:
I added the new line to all of the functions and aggregators files when I changed eval_expr to value. I don't really care as it doesn't affect the docs rendering in html. I was just trying to be consistent.

jbloom22 commented on python/hail/expr/expressions/expression_utils.py:
```diff
@@ -226,3 +226,11 @@ def check_entry_indexed(caller, expr):
+@typecheck(caller=str,
+           expr=Expression)
+def check_row_indexed(caller, expr):
+    if expr._indices != expr._indices.source._row_indices:
```
Good point, I've replaced with `analyze('locus_windows', locus_expr, locus_expr._indices.source._row_indices)`, etc

jbloom22 replied:
This was complicated because the expression may not have a source (which we want to catch with a good error message) and because analyze considers global expressions to be valid row expressions, which we do not want to be the case here. So instead I've fixed this function as well as `check_entry_indexed` to have the right behavior even when no source is present.

jbloom22 commented on python/hail/methods/misc.py:
```diff
@@ -358,4 +368,213 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+        else:
+            raise ValueError(f"array_windows: 'a' must be an ndarray of signed integer or float values, "
+                             f"found dtype {str(a.dtype)}")
+    if (not np.all(a[:-1] <= a[1:])) or (a.size == 1 and np.isnan(a[0])):
```
I don't need to check every element for `nan` because `x <= y` evaluates to false if either is `nan`.  Yeah, it's more precise to say`non-decreasing with respect to index`. This means that `i < j => a(i) <= a(j)`, as opposed to `increasing` which means `i < j => a(i) < a(j)`. I've changed it to ascending order here.

jbloom22 commented on python/hail/methods/misc.py:
```diff
@@ -358,4 +368,213 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+    cum_len_iter = iter(contig_cum_len)
+    i = 0
+    cum_len = next(cum_len_iter)
+    while i < n_loci:
```
no, good point

jbloom22 commented on python/hail/methods/statgen.py:
```diff
@@ -3028,12 +3034,12 @@ def ld_prune(call_expr, r2=0.2, bp_window_size=1000000, memory_per_core=256, kee
 
-    entries = (std_gt_bm @ std_gt_bm.T)._filtered_entries_table(locally_pruned_positions, bp_window_size, False)
-    entries = entries.filter((entries.entry ** 2 >= r2) & (entries.i < entries.j))
+    entries = r2_bm.sparsify_row_intervals(range(0, stops.size), stops, blocks_only=True).entries()
```
it's because in the entries table, we only want to include the edge between `i` and `j` once. We do this by enforcing that `i < j`, which corresponds to the upper triangle. In other words, `r2_bm` is a symmetric matrix and if `i < j` and the entry is at least the threshold, then we use `(i, j)` to represent the edge, not `(j, i)`.

jbloom22 commented on python/hail/tests/test_methods.py:
```diff
@@ -1360,26 +1362,34 @@ def test_ld_prune(self):
     def test_ld_prune_no_prune(self):
-        ds = hl.balding_nichols_model(n_populations=1, n_samples=10, n_variants=100)
-        pruned_table = hl.ld_prune(ds.GT, r2=0.1, bp_window_size=0)
+        ds = hl.balding_nichols_model(n_populations=1, n_samples=10, n_variants=10)
```
Good idea

jbloom22 commented on python/hail/methods/misc.py:
```diff
@@ -358,4 +368,213 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+    if n_loci == 0:
+        return np.zeros(shape=0, dtype=np.int64), np.zeros(shape=0, dtype=np.int64)
+
+    if not np.all(global_pos[:-1] <= global_pos[1:]):
```
I moved the check in `locus_windows` into the for loop. I left the check in `array_windows` as separate since it's more complicated to combine and takes a few microseconds even for millions of elements.

jbloom22 commented:
I did some squashing, all the changes since your review are in the last commit.

jbloom22 commented:
@jigold poke on re-review, I addressed your commends in 667ff7e, created and moved the functions to linalg/utils in 3fdeaa1, and broke out the bit that finds the contig start indices into its own function in ec3a50f so I could test its behavior more rigorously.

@tpoterba poking again on whether you're happy with where I moved the command (see above)

tpoterba commented:
seems fine as `hl.linalg.array_windows`

tpoterba commented:
also fine as `hl.linalg.utils.*`. My concern is resolved

jigold commented on python/test/test_hail/linalg/test_linalg.py:
```diff
@@ -730,3 +730,10 @@ def assert_eq(a, b):
+            global_pos=[0, 0, 1, 2, 3, 4, 5, 5],
+            contig_cum_len=[0, 1, 1, 3, 5, 6, 7])
+
+        self.assertEqual(res, [0, 0, 2, 2, 4, 6])
```
I'm having a hard time reasoning about this test (also the _compute_contig_start_idx function itself). Maybe it's because you have contig lengths with length == 0? Can you add a test where that doesn't occur?

jbloom22 replied:
Yeah, I made this robust to the case of contigs of length 0, even if that's unlikely to occur in practice. The value of the start index of each contig is the index of the first value in `global_pos` that is greater than or equal the cumulative length of the preceding contigs. Length 0 contigs at the end, if present, are dropped.

jigold commented on python/hail/methods/misc.py:
```diff
@@ -467,6 +468,7 @@ def locus_windows(locus_expr, radius, coord_expr=None):
     ...         {'locus': hl.Locus('2', 1), 'cm': 2.0},
     ...         {'locus': hl.Locus('2', 1), 'cm': 2.0},
     ...         {'locus': hl.Locus('3', 3), 'cm': 5.0}]
+    >>>
```
Can you delete this line and just have an empty new line?
